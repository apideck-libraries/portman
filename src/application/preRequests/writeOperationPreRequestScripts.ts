import fs from 'fs'
import { Event, EventDefinition, EventList, Script, ScriptDefinition } from 'postman-collection'
import { PostmanMappedOperation } from 'src/postman'

export const writeOperationPreRequestScripts = (
  pmOperations: PostmanMappedOperation[],
  scripts: string[]
): PostmanMappedOperation[] => {
  return pmOperations.map(pmOperation => {
    const operation = pmOperation.item
    const operationEvents: EventList = operation?.events || []

    let preRequestEvent: EventDefinition =
      operationEvents && operationEvents.find(e => e?.listen === 'prerequest', null)

    if (!preRequestEvent) {
      preRequestEvent = {
        listen: 'prerequest',
        script: {
          exec: [],
          type: 'text/javascript'
        }
      } as EventDefinition
    }

    const scriptContents = scripts.map(src => {
      if (src.startsWith('file:')) {
        return getScriptContent(src.replace('file:', ''))
      } else {
        return src
      }
    })
    const script = new Script(preRequestEvent.script as ScriptDefinition)
    if (script.exec === undefined) script.exec = []
    const exec =
      script.exec && Array.isArray(script.exec)
        ? ([] as string[]).concat(Array.from(script.exec), Array.from(scriptContents))
        : ([script.exec] as string[]).concat(Array.from(scriptContents))
    script.update({ exec: exec.filter(i => Boolean(i)) as string[] })
    preRequestEvent.script = script.toJSON()
    operation.events.add(new Event(preRequestEvent))

    return pmOperation
  })
}

function getScriptContent(scriptPath: string): string {
  try {
    return fs.readFileSync(scriptPath, { encoding:'utf8', flag:'r' })
  } catch(ex) {
    console.error(
      '\x1b[31m', 
      `Config pre-request script file error - no such file or directory "${scriptPath}"`)
    process.exit(1)
  }
}
