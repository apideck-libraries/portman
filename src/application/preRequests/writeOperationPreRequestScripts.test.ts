import { getPostmanMappedOperation } from '../../../__tests__/testUtils/getPostmanMappedOperation'
import { writeOperationPreRequestScripts } from './writeOperationPreRequestScripts'

describe('writeOperationPreRequestScripts', () => {
  it('should inject the configured PreRequestScripts on collection', async () => {
    const pmOperation = await getPostmanMappedOperation()

    const operationPreRequestScripts = [
      'pm.request.headers.add({key: "header_name", value: "header_value" });\n',
      'pm.collectionVariables.set("applicationId", pm.iterationData.get("applicationId") || "1111");\n'
    ]
    writeOperationPreRequestScripts([pmOperation], operationPreRequestScripts)
    expect(pmOperation.item.events.map(({ script }) => script['exec'])).toMatchSnapshot()
  })

  it('should inject the configured PreRequestScripts from file on collection', async () => {
    const pmOperation = await getPostmanMappedOperation()

    const operationPreRequestScripts = [
      'file:src/application/preRequests/__fixtures__/samplePreRequest1.js',
      'file:src/application/preRequests/__fixtures__/samplePreRequest2.js'
    ]
    writeOperationPreRequestScripts([pmOperation], operationPreRequestScripts)
    expect(pmOperation.item.events.map(({ script }) => script['exec'])).toMatchSnapshot()
  })
})
