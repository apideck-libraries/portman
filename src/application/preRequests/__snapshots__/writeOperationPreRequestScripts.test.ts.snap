// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writeOperationPreRequestScripts should inject the configured PreRequestScripts from file on collection 1`] = `
Array [
  Array [
    "pm.request.headers.add({key: \\"header_name\\", value: \\"header_value\\" });
",
    "pm.collectionVariables.set(\\"applicationId\\", pm.iterationData.get(\\"applicationId\\") || \\"1111\\");
",
  ],
]
`;

exports[`writeOperationPreRequestScripts should inject the configured PreRequestScripts on collection 1`] = `
Array [
  Array [
    "pm.request.headers.add({key: \\"header_name\\", value: \\"header_value\\" });
",
    "pm.collectionVariables.set(\\"applicationId\\", pm.iterationData.get(\\"applicationId\\") || \\"1111\\");
",
  ],
]
`;
